scalanet/src/io/iohk/scalanet/peergroup/UDPPeerGroup.scala

```diff
@@ -219,7 +224,10 @@ class UDPPeerGroup[M](val config: Config)(
       }
     }
 
-    override def in: ConnectableObservable[ChannelEvent[M]] = messageSubject
+    //TODO how to finish stream from queue?
+    override def in: Observable[ChannelEvent[M]] = {
+      Observable.repeatEvalF(messageQueue.poll)
```

Comment:
To be clear: because the queue is single-producer-multiple-consumer, all subscribers get every message, right?

Reply:
Ah, no, SPMC just means that there are multiple consumers vying for the same messages, like a worker queue, so a 2nd subscription to in will actually steal the messages from the 1st, no?

Reply:
SPMC here means, that where will be at most one writer thread to that queue, which is guaranted by netty threading model that each channel uses only one thread, and there can be many reader threads, as user can use this observable in many ways and reads its data from many threads.